File: android/two/app/src/main/java/z3d/res/GroupRes.java
package z3d.res;



import android.util.Log;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

import z3d.base.ByteArray;

import z3d.base.CallBackFun;
import z3d.base.GroupBackFun;
import z3d.base.GroupItem;
import z3d.base.RoleBackFun;
import z3d.units.LoadBackFun;
import z3d.units.LoadManager;

public class GroupRes extends BaseRes {

    public List<GroupItem> dataAry ;
    private GroupBackFun groupBackFun;
    private  String TAG="baseres";
    public  void  load(String url,final CallBackFun backFun)
    {
        LoadManager.getInstance().loadUrl(url,LoadManager.BYTE_TYPE, new LoadBackFun() {
            @Override
            public void bfun(HashMap dic) {
                if(dic!=null){

                    ByteArray temp=(ByteArray)dic.get("byte");
                    Log.d(TAG, "bfun");

                    loadComplete((ByteArray) dic.get("byte"), new GroupBackFun() {
                        @Override
                        public void Bfun(GroupRes value) {
                            backFun.StateChange(true);
                        }
                    });



                }else{
                    Log.d(TAG, "bfun: 角色地址错误");
                }


            }
        },null);

    }
    public void  loadComplete(ByteArray buff, GroupBackFun bfun)
    {

        this.groupBackFun=bfun;
        this._byte =buff;

        this.version = this._byte.readInt();
        Log.d("GroupRes ->",   this.version+"===> " );
        this.read(new CallBackFun() {
            @Override
            public void StateChange(boolean State) {
                readNext();
            }
        });//img
    }

    private void   readNext() {
        this.read();//obj
        this.read();//material
        this.read();//particle;

        boolean isGroup = this._byte.readBoolean();

        if (isGroup) {
            int length = this._byte.readInt();
            for (int i = 0; i < length; i++) {
                this.readItem(true);
            }
        } else {
            this.readItem(false);

        }
        this.groupBackFun.Bfun(this);
    }
    private void readItem(boolean isG)
    {
        this.dataAry=new ArrayList<>();
        int types = this._byte.readInt();
        GroupItem item = new GroupItem();
        item.isGroup = isG;

        if (isG) {
            item.x = this._byte.readFloat();
            item.y = this._byte.readFloat();
            item.z = this._byte.readFloat();
            item.scaleX = this._byte.readFloat();
            item.scaleY = this._byte.readFloat();
            item.scaleZ = this._byte.readFloat();
            item.rotationX = this._byte.readFloat();
            item.rotationY = this._byte.readFloat();
            item.rotationZ = this._byte.readFloat();
        }
        if (types == BaseRes.PREFAB_TYPE) {
            item.objUrl = this._byte.readUTF();
            item.materialUrl = this._byte.readUTF();
            item.materialInfoArr = this.readMaterialInfo();
            item.types = BaseRes.PREFAB_TYPE;

        } else if (types == BaseRes.SCENE_PARTICLE_TYPE) {
            item.particleUrl = this._byte.readUTF();
            item.types = BaseRes.SCENE_PARTICLE_TYPE;
        }

        this.dataAry .add(item);

    }
}

File: new/src/Pan3d/res/GroupRes.ts
module Pan3d {
    export class GroupRes extends BaseRes {
        private _fun: Function;

        public dataAry: Array<GroupItem>;

        private _objDic: Object;
        private _particleDic: Object;
        private _materialDic: Object;

        public load(url: string, $fun: Function): void {

            this._fun = $fun;
            LoadManager.getInstance().load(url, LoadManager.BYTE_TYPE, ($byte: ArrayBuffer) => {
                this.loadComplete($byte);
            });
        }

        public loadComplete($byte: ArrayBuffer): void {
            this.dataAry = new Array;

            this._byte = new Pan3dByteArray($byte);
            this._byte.position = 0

            this.version = this._byte.readInt();

            this.read(() => { this.readNext() });//img


        }

        private readNext(): void {
            this.read();//obj
            this.read();//material
            this.read();//particle;

            var isGroup: boolean = this._byte.readBoolean();

            if (isGroup) {
                var length: number = this._byte.readInt();

                for (var i: number = 0; i < length; i++) {
                    this.readItem(true);
                }

            } else {
                this.readItem(false);

            }

            this._fun();

            this._fun = null;
            this._byte = null;
        }

        private readItem(isG: boolean): void {
            var types: number = this._byte.readInt();
            var item: GroupItem = new GroupItem();
            item.isGroup = isG;

            if (isG) {
                item.x = this._byte.readFloat();
                item.y = this._byte.readFloat();
                item.z = this._byte.readFloat();
                item.scaleX = this._byte.readFloat();
                item.scaleY = this._byte.readFloat();
                item.scaleZ = this._byte.readFloat();
                item.rotationX = this._byte.readFloat();
                item.rotationY = this._byte.readFloat();
                item.rotationZ = this._byte.readFloat();
            }

            if (types == BaseRes.PREFAB_TYPE) {
                item.objUrl = this._byte.readUTF();
                item.materialUrl = this._byte.readUTF();

                if (this.version >= 4) {
                    item.materialInfoArr = this.readMaterialInfo();
                }
                item.types = BaseRes.PREFAB_TYPE;

            } else if (types == BaseRes.SCENE_PARTICLE_TYPE) {
                item.particleUrl = this._byte.readUTF();
                item.types = BaseRes.SCENE_PARTICLE_TYPE;
            }

            this.dataAry.push(item);
        }


        public initReg(): void {
            this._objDic = new Object;
            this._materialDic = new Object;
            this._particleDic = new Object;

            for (var i: number = 0; i < this.dataAry.length; i++) {
                var item: GroupItem = this.dataAry[i];
                if (item.objUrl) {
                    this._objDic[this.scene3D.fileRoot + item.objUrl] = true;
                }
                if (item.materialUrl) {
                    this._materialDic[this.scene3D.fileRoot + item.materialUrl] = true;
                }
                if (item.particleUrl) {
                    this._particleDic[this.scene3D.fileRoot + item.particleUrl] = true;
                }
            }

            for (var key in this._objDic) {
                this.scene3D.objDataManager.registerUrl(key);
            }

            for (var key in this._materialDic) {
                this.scene3D.materialManager.registerUrl(key);
            }

            for (var key in this._particleDic) {
                this.scene3D.particleManager.registerUrl(key);
            }

        }
 



    }

  
}